crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2845,14 +2846,218 @@ impl<'db> Type<'db> {
+                            Parameter::new(
+                                Some(Name::new_static("contravariant")),
+                                Some(KnownClass::Bool.to_instance(db)),
+                                ParameterKind::KeywordOnly {
+                                    default_ty: Some(Type::BooleanLiteral(false)),
+                                },
+                            ),
+                        ]),
+                        Some(KnownClass::TypeVar.to_instance(db)),
+                    )])
+                }
+
+                let mut binding = bind_call(db, arguments, overloads(db), self);
+                if binding.matching_overload_mut().is_none() {
+                    return Err(CallError::BindingError { binding });
+                };
+
                 binding.into_outcome()
             }
 
+            // TODO: currently we can't use typeshed to infer calls using generic
+            // logic below due to a salsa query cycle. When this is fixed, we should
+            // this arm and aboid special casing `Object` in the `ClassLiteral` arm.
+            Type::ClassLiteral(ClassLiteralType { class })
+                if class.is_known(db, KnownClass::Object) =>
+            {
+                // ```py
+                // class object:
+                //    def __init__(self) -> None: ...
+                //    def __new__(cls) -> Self: ...
+                // ```
+                #[salsa::tracked(return_ref)]
+                fn overloads<'db>(db: &'db dyn Db) -> CallableSignature<'db> {
+                    CallableSignature::from_overloads([Signature::new(
+                        Parameters::new([]),
+                        Some(KnownClass::Object.to_instance(db)),
+                    )])
+                }
+
+                let mut binding = bind_call(db, arguments, overloads(db), self);
+                if binding.matching_overload_mut().is_none() {
+                    return Err(CallError::BindingError { binding });
+                };
+
+                binding.into_outcome()
+            }
+
+            // TODO annotated return type on `__new__` or metaclass `__call__`
+            Type::ClassLiteral(ClassLiteralType { class }) => {
+                // User defined or known class not handled by the above cases
+                let instance_ty = Type::Instance(InstanceType { class });
+
+                // As of now we do not model custom `__call__` on meta-classes, so the code below
+                // only deals with interplay between `__new__` and `__init__` methods.
+                // The logic is roughly as follows:
+                // 1. If `__new__` is defined anywhere in the MRO (except for `object`, since it is always
+                //    present), we call it, analyze outcome and stop. There is no need to try binding
+                //    `__init__` in this case, since any signature mismatches should be handled at definition
+                //    site of the class.
```

Comment:
I realized this isn't right, as mentioned in a comment above. We can't really check `__init__` vs `__new__` compat at class definition, because they are allowed to differ; the effective constructor signature is the intersection of the two (whatever calls satisfy both signatures.)

I think if `__new__` exists and binding to it fails, it's fine to stop there. If you have a similar `__new__` and `__init__` signature, and you just pass the wrong arguments, we don't want two diagnostics about it. But if `__new__` exists and binding to it succeeds, we need to go ahead and try `__init__` as well and make sure that succeeds too.

Sorry I led you in the wrong direction here earlier.